### PR TITLE
Feature/vue frontend

### LINKED ISSUE #3: Front-End
To build a user interface. Possibly some buttons included such as "How to play?", "Start", etc.

Upon starting the game, a text editor with some pre-defined, broken, code appears (sent from the backend). Perhaps have a timer somewhere on the screen to record how fast they complete the debugging process.

- [x] First layer - interactive GUI text editor
  - [x] Buttons for game details, starting, level of difficulty, join other games(?)
- [ ] Second layer - websocket integration for live publication
  - [ ] Set-up the connection when the first layer is in view
  - [ ] Receive data from server as a code template for the user to debug
  - [ ] Send data (code snapshots) every x seconds (depending on difficulty selected)

Will probably want to separate out each larger task into different PRs.

Any other elements can be discussed below.

### LINKED ISSUE #53: Code Timer
We need a timer that monitors the time it takes the user to solve the challenge.

We mainly have two options:
- Implement it in the front end, but there is a risk that the user will mess with it.
- Implement it in the back end, the timer would run as a background task.

Any changes can be discussed below.

### PR DESCRIPTION
**Pull request by eklipse18**
Merge the frontend implemented in Vue.
Todo:
- Clean up UI and the code.

### CODE REVIEW

**Review by LukeHankey — COMMENTED** [review]
Looks good. I personally prefer the previous theme but I don't think choosing the theme is anything major at this point :p

**Comment by eklipse18** [comment]
I just need to handle empty room codes now

**Comment by bensgilbert** [comment]
There needs to be a way to enter the username.

**Comment by bensgilbert** [comment]
Not ready yet, I need to finish up the web socket JS :)

File: frontend/src/components/Home.vue
Comment (LukeHankey) [inline frontend/src/components/Home.vue]: Can you default the difficulty to 1

**Comment by LukeHankey** [comment]
Front end looks really nice. Can we have the collaborators section show the username as in "Luke - You" as other people might have the same username.

Perhaps connect the websocket in a different PR. This looks good to get the frontend setup visually. Adding logic would be in a different scope

**Comment by eklipse18** [comment]
> Perhaps connect the websocket in a different PR. This looks good to get the frontend setup visually. Adding logic would be in a different scope

Yeah, I agree. I think let's merge this PR after I make these couple fixes instead of waiting for client JS to be finished.

**Review by LukeHankey — APPROVED** [review]
Looks good. ~Think you just need to run prettier on the files to pass the GHA.~ Seems prettier runs automatically by the bot. Just the few fixes on DeepSource to pass then

**Comment by eklipse18** [comment]
Do we need to satisfy deepsource before merging?

**Comment by LukeHankey** [comment]
> Do we need to satisfy deepsource before merging?

Ideally, yes. It gives some really useful tips for better code quality

**Comment by eklipse18** [comment]
@Evorage0 do you mind taking up the issues on deepsource? because most of it is from your part I think and I don't want to meddle with it because it may have unwanted consequences.

File: frontend/src/assets/js/main.js
Comment (LukeHankey) [inline frontend/src/assets/js/main.js]: Might still need this as the `window.handleContentChange` requires the websocket.

File: frontend/src/assets/js/main.js
Comment (LukeHankey) [inline frontend/src/assets/js/main.js]: `const list = ...`

File: frontend/src/assets/js/main.js
Comment (LukeHankey) [inline frontend/src/assets/js/main.js]: Move line 91 above the switch statement and declare it as const. Can remove line 80 with this as well.

File: frontend/src/assets/js/main.js
Comment (LukeHankey) [inline frontend/src/assets/js/main.js]: Commenting here because I can't comment on line 81. Also move `var elem = ...` to above the switch statement and declare this one as `let`. It is only used in the connect and disconnect so they shouldn't interfere with each other ( I think)

Comment (kappq) [reply frontend/src/assets/js/main.js]: However, we can actually get the element only when it's created (basically when the connect event is sent).

Comment (LukeHankey) [reply frontend/src/assets/js/main.js]: Guess this makes sense within it's scope as you don't want to be creating the list element on each request

File: frontend/src/components/Room.vue
Comment (LukeHankey) [inline frontend/src/components/Room.vue]: `convert`

File: server/events.py
Comment (LukeHankey) [inline server/events.py]: `user (optional):`

File: server/event_handler.py
Comment (LukeHankey) [inline server/event_handler.py]: Why is it called `connect_data`?

File: server/event_handler.py
Comment (LukeHankey) [inline server/event_handler.py]: The cast here is not needed as the above `connect_data` is the same thing.

File: server/events.py
Comment (LukeHankey) [inline server/events.py:80]: `collaborators (optional):`
`time (optional):`

**Review by LukeHankey — CHANGES_REQUESTED** [review]
Just comments left to review

File: server/events.py
Comment (LukeHankey) [inline server/events.py:81]: Why make it optional? This data could be sent with all sync responses and requests, no? Same with all of these optional fields.

**Review by LukeHankey — APPROVED** [review]

**Pull request merged**
